### PR TITLE
APEXMALHAR-2248 Added SpillableSet and SpillableSetMultimap implementation

### PR DESCRIPTION
**Pull request by davidyan74**
@tweise @siyuanh @ilooner please review

### CODE REVIEW

**Comment by siyuanh** [comment]
Overall looks good. But I have some objection here about the design especially the class hierarchy is apparently over-designed to me. I think the intermediate interface is useless. For example interface SpillableArrayList just extends List. It has no method, and if it is used as label, we already have Spillable and SpillableComponent.  What I'm thinking it cut off this layer, so the SpillableArrayListImpl would directly implements SpillableComponent and List.

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableSetMultimapImpl.java [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+@DefaultSerializer(FieldSerializer.class)
+@InterfaceStability.Evolving
+public class SpillableSetMultimapImpl<K, V> implements Spillable.SpillableSetMultimap<K, V>,
+    Spillable.SpillableComponent
+{
+  public static final int DEFAULT_BATCH_SIZE = 1000;
+  public static final byte[] SIZE_KEY_SUFFIX = new byte[]{(byte)0, (byte)0, (byte)0};
+
+  private transient WindowBoundedMapCache<K, SpillableSetImpl<V>> cache = new WindowBoundedMapCache<>();
+  private transient boolean isRunning = false;
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableSetMultimapImpl.java]: What are isRunning and isInWindow for?

File: library/src/main/java/org/apache/apex/malhar/lib/utils/serde/SerdeCollectionSlice.java [hunk 2]
```diff
@@ -37,35 +37,40 @@
  * @since 3.5.0
  */
 @InterfaceStability.Evolving
-public class SerdeListSlice<T> implements Serde<List<T>, Slice>
+public class SerdeCollectionSlice<T, CollectionT extends Collection<T>> implements Serde<CollectionT, Slice>
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/utils/serde/SerdeCollectionSlice.java:40]: Is it better to have Collection<T> in Serde?

File: library/src/main/java/org/apache/apex/malhar/lib/utils/serde/SerdeCollectionSlice.java [hunk 3]
```diff
@@ -37,35 +37,40 @@
...
   {
     this.serde = Preconditions.checkNotNull(serde);
+    this.collectionClass = Preconditions.checkNotNull(collectionClass);
   }
 
   @Override
-  public Slice serialize(List<T> objects)
+  public Slice serialize(CollectionT objects)
   {
     Slice[] slices = new Slice[objects.size()];
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/utils/serde/SerdeCollectionSlice.java:66]: Probably not in this PR. But I'm thinking about some special class like ArraySlice which holds buffer and array of offsets in one object.  But I don't know how it will fit into this method.

Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/utils/serde/SerdeCollectionSlice.java:66]: But I think this should be useful in many places

**Comment by otterc** [comment]
@siyuanh dis-agree with the comment on Spillable class hierarchy. It is not over-designed. Instead it makes it easier for developer to find out all the different implementations for a specific datastructure.

We can have multiple SpillableList implementations. While writing an operator which needs a spillable list, it is easier for me to find all the existing implementations  that are of type SpillableList.

**Comment by ilooner** [comment]
@siyuanh a benefit to having the interfaces that way is that we can easily add methods to a SpillableDataStructure signature. Yes there are no methods in any of those interfaces now, but there may be different methods added for different datastructures in the future. Furthermore it is useful to document the contract of each datastructure and all its implementations, which may differ from their typical in memory cousins. Ultimately I think it makes future modification and documentation easier, while preserving readability.

File: library/src/main/java/org/apache/apex/malhar/lib/utils/serde/SerdeCollectionSlice.java (hunk 2 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/utils/serde/SerdeCollectionSlice.java:40]: The serde is for any collection, not just a List. That's why I made this change, since I need a serde for a Set and the same serde works for a Set as well.

**Comment by siyuanh** [comment]
@ilooner I agree we need a interface for all/or at least a group of common Spillable data structures. It like in Java we have List/ OrderedList interface but we don't have ArrayList and TreeList interface. You can also argue they have different methods internally, but that wouldn't be in the interface level. Interface delivers another level of abstraction in my mind and my understanding of java library. To the name  SpillableArrayList and SpillableByteArrayListMultimap are already very very specific I don't see any place would have something called ArrayList interface and/or ByteArrayListMultimap interface.

**Comment by siyuanh** [comment]
@chandnisingh Your point is a voting question, whether you think it's easier or I think it's easier doesn't mean anything. Let's discuss this in mailing list. And actually SpillableList is easier to understand instead of SpillableArrayList interface.

**Comment by otterc** [comment]
@siyuanh well than this holds for your initial comment as well 😀.  It is just your opinion that it is "overdesigned"  which in your words doesn't mean anything.

**Comment by tweise** [comment]
Perhaps the discussion about interface hierarchy can be separated from this PR. There are different perspectives. Another consideration is that you can start with something that is simple and meets the current needs and can be extended with use cases and learning from it.

**Comment by davidyan74** [comment]
@ilooner I tried to implement `SpillableSetMultiMap.removeAll(K key)` and I'm having some trouble doing it. I'm making this PR review-only and will push my current implementation of removeAll later today and would love to get your feedback. Also, I am curious on your reasons behind the word "Byte" in all the class names. Should I rename SpillableSet and SpillableSetMultimap interfaces and their implementation to say SpillableByteSet, etc?

**Comment by ilooner** [comment]
@davidyan74 I would recommend a lazy delete for now, by setting an empty byte array as the size of the the set for a key in SpillableSetMultimap. The data will still be in managed state, until managed state expires it. If new data is set for the same key later, it is fine, it will just overwrite the old keys. Just be sure that in SpillableSet you point the last node in your linked list to an empty byte array, and recognize an empty byte array and null as an indicator of nonexistence for a value throughout your code. 

I think this is a good compromise because it makes removals fast, and leverage's managed state's efficient purging mechanism to prevent unbounded growth of data.

**Comment by davidyan74** [comment]
@ilooner I have fixed several potential bugs.
Please take a look at the last 3 commits in this PR and see whether or not you agree with them.
The recovery unit test is still failing. I'll take a deeper look at what's going on tomorrow but if you know what's going on in there, please let me know.

File: library/src/test/java/org/apache/apex/malhar/lib/state/spillable/SpillableSetMultimapImplTest.java [hunk 4]
```diff
@@ -185,24 +205,8 @@ public long simpleMultiKeyTestHelper(SpillableStateStore store,
...
-
-    Assert.assertFalse(map.containsKey(key));
-
-    SpillableTestUtils.checkValue(store, 0L,
-        SliceUtils.concatenate(keyBytes, SpillableSetMultimapImpl.SIZE_KEY_SUFFIX), 0, 0, serdeInt);
-
-
     map.endWindow();
     store.endWindow();
-    store.beforeCheckpoint(nextWindowId);
```
Comment (ilooner) [inline library/src/test/java/org/apache/apex/malhar/lib/state/spillable/SpillableSetMultimapImplTest.java]: Why was this removed? In order to correctly simulate a checkpoint with managed state you need to beforeCheckpoint and then checkpointed. It looks like there may be a pre existing issue on line 228 because the window is incremented before checkpointed is called... My bad.

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableSetMultimapImpl.java [hunk 5]
```diff
@@ -64,6 +64,8 @@
   private long bucket;
   private Serde<K, Slice> serdeKey;
   private Serde<V, Slice> serdeValue;
+  private int sequence = 0;
```
Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableSetMultimapImpl.java]: I would feel better if this was a long.

**Comment by ilooner** [comment]
@davidyan74 left a couple comments. overall lgtm

**Comment by ilooner** [comment]
@davidyan74 just realize we need to save the current sequence for a key, if the Set for a key is requested multiple times (like on recovery) then the set may be assigned a different sequence number than the first time.

**Comment by davidyan74** [comment]
@ilooner  If a new set is assigned a different sequence number, what would be the harm? And even there is harm, if the tuples are coming in in an idempotent way, the sequence number should remain the same, right?

**Comment by ilooner** [comment]
@davidyan74 consider the following scenario.
1. You request key A and the corresponding set is assigned a sequence number of 20. 
2. Then lets say you request 80 more different keys, so your sequence number is now 100. 
3. Then you checkpoint
4. Then you fail and get restored to the checkpoint
5. At this point your cache of SpillableSets is empty because it is transient and your sequence number is 100
6. Now we request key A, so a new Spillable set is made with sequence number 101 for key A. But wait, the first time we requested key A the sequence number was 20! So now our new key prefix is wrong and we have lost all the data we stored for that key.

A similar thing can happen in the case when a SpillableSet gets evicted from the cache and gets brought back

**Comment by davidyan74** [comment]
@ilooner I see what you're saying. That would mean a sequence number is useless in this case. I think I just need to have a marker for each spillableset that marks whether it has been removed. Or do you have other ideas?

**Comment by davidyan74** [comment]
@ilooner i think what you said was the reason why the recovery test was failing.

**Comment by ilooner** [comment]
@davidyan74 I think the marker would work

**Comment by davidyan74** [comment]
@ilooner @chandnisingh How does Managed State decide to purge with the current Spillable data structures? The current `remove` implementation of spillable data structures seems to only do a marking (by setting the value to be EMPTY_BYTE_ARRAY) and does not do explicit remove, and Managed State has no way to know that they are actually no longer valid. Does that mean the data would just grow indefinitely?

**Comment by davidyan74** [comment]
This PR depends on #416 #417 #418. Will issue a rebased PR once they are all merged.

**Comment by davidyan74** [comment]
@siyuanh please review and merge

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableSetImpl.java [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+    {
+      this.valid = valid;
+      this.next = next;
+    }
+
+    boolean valid;
+    T next;
+  }
+
+  public static class SerdeListNodeSlice<T> implements Serde<ListNode<T>, Slice>
```
Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableSetImpl.java]: ListNodeSliceSerde is a better name

Comment (siyuanh) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableSetImpl.java]: Follow the convention. I see never mind
